doc/man/man3/MappedROUKF.3
.TH "MappedROUKF" 3 "Fri Mar 23 2018" "Version 1.0" "A Kalman Library (AKL)" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MappedROUKF
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <MappedROUKF\&.h>\fP
.PP
Inherits \fBAbstractROUKF\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBMAPPING_TYPE\fP { \fBDEFAULT\fP, \fBPOSITIVE\fP, \fBRANGED\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMappedROUKF\fP (int \fBnObservations\fP, int \fBnStates\fP, int \fBnParameters\fP, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution)"
.br
.ti -1c
.RI "\fBMappedROUKF\fP (int \fBnObservations\fP, int \fBnStates\fP, int \fBnParameters\fP, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution, \fBMappedROUKF::MAPPING_TYPE\fP mappingType, vector< double > mappingParameters)"
.br
.ti -1c
.RI "\fBMappedROUKF\fP (int \fBnObservations\fP, int \fBnStates\fP, int \fBnParameters\fP, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution, \fBCompositeParameterMapper\fP *\fBmapper\fP)"
.br
.ti -1c
.RI "\fB~MappedROUKF\fP ()"
.br
.ti -1c
.RI "double \fBexecuteStep\fP (vector< double > Zkhatc, forwardOp A, observationOp H)"
.br
.ti -1c
.RI "double \fBexecuteStepParallel\fP (vector< double > Zkhatc, forwardOp A, observationOp H, int seed, MPI_Comm local_comm, MPI_Comm masters_comm)"
.br
.ti -1c
.RI "void \fBreset\fP (int \fBnObservations\fP, int \fBnStates\fP, int \fBnParameters\fP, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution)"
.br
.ti -1c
.RI "void \fBreplaceMapper\fP (\fBCompositeParameterMapper\fP *\fBmapper\fP)"
.br
.ti -1c
.RI "virtual void \fBgetParameters\fP (double **ThetaC) override"
.br
.ti -1c
.RI "virtual void \fBsetParameters\fP (double *ThetaC) override"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBCompositeParameterMapper\fP * \fBmapper\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
An example for the usage of the library:
.PP
.PP
.nf
   int (*ptA)(double*, int) = NULL;
void (*ptH)(double*, int, double*, int) = NULL;
ptA = &forwardOpElasticity;
ptH = &observerOpElasticity;

 //    Initialization parameters
for (int i = 0; i < nParameters; i++) {
   initialGuess[i] = 2\&.5E5;            // -> X0
   parametersUncertainties[i] =1E8;    // -> U0^{-1}
}
ROUKF *kalmanInstance = new ROUKF(nStates, nParameters, statesUncertainties,
parametersUncertainties);

// Set initial condition
kalmanInstance->setState(initialGuess, nParameters);

for (int it = 0; it < 3000; it++) {
   error = kalmanInstance->executeStep(observation, nStates, ptA, ptH);
}

// Get the Kalman estimation
kalmanInstance->getState(&sol);            // -> XSol
.fi
.PP
 Class that implements the reduced order unscented Kalman filter\&. 
.PP
Definition at line 57 of file MappedROUKF\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBMappedROUKF::MAPPING_TYPE\fP"
Reparametrization type\&. 
.PP
Definition at line 64 of file MappedROUKF\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MappedROUKF::MappedROUKF (int nObservations, int nStates, int nParameters, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution)"
Creates the covariance matrixes and sigma points associated with the extended vector X and their uncertainty\&. 
.PP
\fBParameters:\fP
.RS 4
\fInObservations\fP Quantity of observations\&. 
.br
\fInStates\fP Quantity of states\&. 
.br
\fInParameters\fP Quantity of parameters\&. 
.br
\fIobservationsUncertainty\fP Vector with the uncertainty of each state in X\&. 
.br
\fIparametersUncertainty\fP Vector with the uncertainty of each parameter in Theta\&. 
.br
\fIsigmaDistribution\fP Type of sigmas applied to assess the unscented transform\&. 
.RE
.PP

.PP
Definition at line 18 of file MappedROUKF\&.cpp\&.
.SS "MappedROUKF::MappedROUKF (int nObservations, int nStates, int nParameters, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution, \fBMappedROUKF::MAPPING_TYPE\fP mappingType, vector< double > mappingParameters)"
Creates the covariance matrixes and sigma points associated with the extended vector X and their uncertainty\&. The chosen mapping is the same for all parameters\&. 
.PP
\fBParameters:\fP
.RS 4
\fInObservations\fP Quantity of observations\&. 
.br
\fInStates\fP Quantity of states\&. 
.br
\fInParameters\fP Quantity of parameters\&. 
.br
\fIobservationsUncertainty\fP Vector with the uncertainty of each state in X\&. 
.br
\fIparametersUncertainty\fP Vector with the uncertainty of each parameter in Theta\&. 
.br
\fIsigmaDistribution\fP Type of sigmas applied to assess the unscented transform\&. 
.br
\fImappingType\fP Type of mapping between the problem and the kalman parameters\&. 
.br
\fImappingParameters\fP Parameters for the chosen mapping type\&. 
.RE
.PP

.PP
Definition at line 52 of file MappedROUKF\&.cpp\&.
.SS "MappedROUKF::MappedROUKF (int nObservations, int nStates, int nParameters, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution, \fBCompositeParameterMapper\fP * mapper)"
Creates the covariance matrixes and sigma points associated with the extended vector X and their uncertainty\&. The mapping between the problem and the kalman parameters is given by the user customized \fCmapper\fP mapping function\&. 
.PP
\fBParameters:\fP
.RS 4
\fInObservations\fP Quantity of observations\&. 
.br
\fInStates\fP Quantity of states\&. 
.br
\fInParameters\fP Quantity of parameters\&. 
.br
\fIobservationsUncertainty\fP Vector with the uncertainty of each state in X\&. 
.br
\fIparametersUncertainty\fP Vector with the uncertainty of each parameter in Theta\&. 
.br
\fIsigmaDistribution\fP Type of sigmas applied to assess the unscented transform\&. 
.br
\fImapper\fP User customized mapping function between the problem and the kalman parameters\&. 
.RE
.PP

.PP
Definition at line 99 of file MappedROUKF\&.cpp\&.
.SS "MappedROUKF::~MappedROUKF ()"
Void destructor\&. 
.PP
Definition at line 129 of file MappedROUKF\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "double MappedROUKF::executeStep (vector< double > Zkhatc, forwardOp A, observationOp H)"
Performs one step of the Kalman filtering process in serial execution of the sigma points\&. 
.PP
\fBParameters:\fP
.RS 4
\fIZkhatc\fP Current observations estimations\&. 
.br
\fIA\fP Forward operator\&. 
.br
\fIH\fP Observation operator; 
.RE
.PP
\fBReturns:\fP
.RS 4
Current L2 norm of the errors across all observations\&. 
.RE
.PP

.PP
Definition at line 132 of file MappedROUKF\&.cpp\&.
.SS "double MappedROUKF::executeStepParallel (vector< double > Zkhatc, forwardOp A, observationOp H, int seed, MPI_Comm local_comm, MPI_Comm masters_comm)"
Performs one step of the Kalman filtering process with parallel execution of the sigma points\&. 
.PP
\fBParameters:\fP
.RS 4
\fIZkhatc\fP Current observations estimations\&. 
.br
\fIA\fP Forward operator\&. 
.br
\fIH\fP Observation operator; 
.br
\fIseed\fP Sigma point ID for the current MPI process\&. 
.br
\fIlocal_comm\fP Communicator of all MPI processes that solve the sigma point \fCseed\fP\&. 
.br
\fImasters_comm\fP Communicator of the master MPI processes of each sigma point \fCseed\fP\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Current L2 norm of the errors across all observations\&. 
.RE
.PP

.PP
Definition at line 203 of file MappedROUKF\&.cpp\&.
.SS "void MappedROUKF::getParameters (double ** ThetaC)\fC [override]\fP, \fC [virtual]\fP"
Return the current parameter estimatives in the array \fCThetaC\fP\&. 
.PP
\fBParameters:\fP
.RS 4
\fIThetaC\fP Current parameter estimatives\&. 
.RE
.PP

.PP
Reimplemented from \fBAbstractROUKF\fP\&.
.PP
Definition at line 313 of file MappedROUKF\&.cpp\&.
.SS "void MappedROUKF::replaceMapper (\fBCompositeParameterMapper\fP * mapper)"
Replaces the current \fCmapper\fP for the new \fBCompositeParameterMapper\fP instance\&. 
.PP
\fBParameters:\fP
.RS 4
\fImapper\fP New mapper instance\&. 
.RE
.PP

.PP
Definition at line 326 of file MappedROUKF\&.cpp\&.
.SS "void MappedROUKF::reset (int nObservations, int nStates, int nParameters, vector< double > observationsUncertainty, vector< double > parametersUncertainty, \fBSigmaPointsGenerator::SIGMA_DISTRIBUTION\fP sigmaDistribution)"
Returns to the initial state of the kalman filter\&. Not fully tested 
.PP
\fBParameters:\fP
.RS 4
\fInObservations\fP Quantity of observations\&. 
.br
\fInStates\fP Quantity of states\&. 
.br
\fInParameters\fP Quantity of parameters\&. 
.br
\fIobservationsUncertainty\fP Vector with the uncertainty of each state in X\&. 
.br
\fIparametersUncertainty\fP Vector with the uncertainty of each parameter in Theta\&. 
.br
\fIsigmaDistribution\fP Type of sigmas applied to assess the unscented transform\&. 
.RE
.PP

.PP
Definition at line 286 of file MappedROUKF\&.cpp\&.
.SS "void MappedROUKF::setParameters (double * ThetaC)\fC [override]\fP, \fC [virtual]\fP"
Sets the initial parameter estimatives\&. 
.PP
\fBParameters:\fP
.RS 4
\fIThetaC\fP Initial parameter estimatives\&. 
.RE
.PP

.PP
Reimplemented from \fBAbstractROUKF\fP\&.
.PP
Definition at line 320 of file MappedROUKF\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBCompositeParameterMapper\fP* MappedROUKF::mapper\fC [private]\fP"
Mapping function between the problem parameters and the kalman parameters\&. 
.PP
Definition at line 60 of file MappedROUKF\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for A Kalman Library (AKL) from the source code\&.


MappedROUKF.cpp
/*
 * ROUKF.cpp
 *
 *  Created on: Oct 26, 2015
 *      Author: Gonzalo D. Maso Talou
 */

#include "iostream"
#include "MappedROUKF.h"
#include "mapping/IdentityParameterMapper.h"
#include "mapping/ExponentialParameterMapper.h"
#include "mapping/SigmoidParameterMapper.h"
#include <cmath>

using namespace arma;
using namespace std;

MappedROUKF::MappedROUKF(int nObservations, int nStates, int nParameters, vector<double> observationsUncertainty, vector<double> parametersUncertainty,
		SigmaPointsGenerator::SIGMA_DISTRIBUTION sigmaDistribution) :
		AbstractROUKF() {

	this->nObservations = nObservations;
	this->nStates = nStates;
	this->nParameters = nParameters;

	X = zeros(nStates, 1);
	Theta = zeros(nParameters, 1);

	LTheta = eye(nParameters, nParameters);
	LX = zeros(nStates, nParameters);
	U = eye(nParameters, nParameters);
	Wi = speye(nObservations, nObservations);

	mat diagParameter(&(parametersUncertainty[0]), 1, nParameters);
	U.diag() = 1. / diagParameter;
	mat diagObservations(&(observationsUncertainty[0]), 1, nObservations);
	Wi.diag() = 1. / diagObservations;

	SigmaPointsGenerator::generateSigmaPoints(nParameters, sigmaDistribution, &sigma);
	alpha = 1. / sigma.n_cols;

	Dsigma = alpha * eye(sigma.n_cols, sigma.n_cols);
	Dsigma = Dsigma * sigma.t();
	Pa = sigma * Dsigma;

	vector<AbstractParameterMapper *> mappers;
	mappers.push_back(new IdentityParameterMapper());
	vector<int> paramsPerMapper = { (int) sigma.n_cols };
	mapper = new CompositeParameterMapper(paramsPerMapper, mappers);
}

MappedROUKF::MappedROUKF(int nObservations, int nStates, int nParameters, vector<double> observationsUncertainty, vector<double> parametersUncertainty,
		SigmaPointsGenerator::SIGMA_DISTRIBUTION sigmaDistribution, MAPPING_TYPE mappingType, vector<double> mappingParameters) :
		AbstractROUKF() {
	this->nObservations = nObservations;
	this->nStates = nStates;
	this->nParameters = nParameters;

	X = zeros(nStates, 1);
	Theta = zeros(nParameters, 1);

	LTheta = eye(nParameters, nParameters);
	LX = zeros(nStates, nParameters);
	U = eye(nParameters, nParameters);
	Wi = speye(nObservations, nObservations);

	mat diagParameter(&(parametersUncertainty[0]), 1, nParameters);
	U.diag() = 1. / diagParameter;
	mat diagObservations(&(observationsUncertainty[0]), 1, nObservations);
	Wi.diag() = 1. / diagObservations;

	SigmaPointsGenerator::generateSigmaPoints(nParameters, sigmaDistribution, &sigma);
	alpha = 1. / sigma.n_cols;

	Dsigma = alpha * eye(sigma.n_cols, sigma.n_cols);
	Dsigma = Dsigma * sigma.t();
	Pa = sigma * Dsigma;

	vector<AbstractParameterMapper *> mappers;
	vector<int> paramsPerMapper = { (int) sigma.n_cols };
	switch (mappingType) {
	case DEFAULT:
		mappers.push_back(new IdentityParameterMapper());
		break;
	case POSITIVE:
		mappers.push_back(new ExponentialParameterMapper());
		break;
	case RANGED:
		mappers.push_back(new SigmoidParameterMapper(mappingParameters[0], mappingParameters[1]));
		break;
	default:
		cout << "Unrecognize type of parameter mapping, default is employed" << endl;
		mappers.push_back(new IdentityParameterMapper());
		break;
	}
	mapper = new CompositeParameterMapper(paramsPerMapper,mappers);
}

MappedROUKF::MappedROUKF(int nObservations, int nStates, int nParameters, vector<double> observationsUncertainty, vector<double> parametersUncertainty,
		SigmaPointsGenerator::SIGMA_DISTRIBUTION sigmaDistribution, CompositeParameterMapper *mapper) :
		AbstractROUKF() {
	this->nObservations = nObservations;
	this->nStates = nStates;
	this->nParameters = nParameters;

	X = zeros(nStates, 1);
	Theta = zeros(nParameters, 1);

	cout << X << endl;
	cout << Theta << endl;

	LTheta = eye(nParameters, nParameters);
	LX = zeros(nStates, nParameters);
	U = eye(nParameters, nParameters);
	Wi = speye(nObservations, nObservations);

	mat diagParameter(&(parametersUncertainty[0]), 1, nParameters);
	U.diag() = 1. / diagParameter;
	mat diagObservations(&(observationsUncertainty[0]), 1, nObservations);
	Wi.diag() = 1. / diagObservations;

	cout << U << endl;
	cout << Wi << endl;

	SigmaPointsGenerator::generateSigmaPoints(nParameters, sigmaDistribution, &sigma);
	alpha = 1. / sigma.n_cols;

	Dsigma = alpha * eye(sigma.n_cols, sigma.n_cols);
	Dsigma = Dsigma * sigma.t();
	Pa = sigma * Dsigma;

	this->mapper = mapper;
}

MappedROUKF::~MappedROUKF(){
}

double MappedROUKF::executeStep(vector<double> zkhatc, forwardOp A, observationOp H) {

	//	Matrixes
	mat Thetak(nParameters, sigma.n_cols), Xk(nStates, sigma.n_cols), Zk(nObservations, sigma.n_cols);
	mat HL;
	mat invU = inv(U);
	mat C = chol(invU);

	//	Column vectors
	mat thetak, xk, zk;
	mat zkhat(&(zkhatc[0]), nObservations, 1);

	double *zkdata;
	zkdata = new double[nObservations];

	for (unsigned int i = 0; i < sigma.n_cols; i++) {
		//	Sampling
		mat s = sigma.col(i);
		xk = X + LX * C.t() * s;
		thetak = Theta + LTheta * C.t() * s;

		//	Transform theta_k -kalman parameters- to problem values -problem parameters-

		//	Propagate sigma point
		//	Copies xk to a primitive type in order to achieve armadillo independence in extern programs
		vector<double> thetakdata(thetak.memptr(), thetak.memptr() + nParameters);
		vector<double> xkdata(xk.memptr(), xk.memptr() + nStates);

		thetakdata = mapper->unmap(thetakdata);
		(*A)(&(xkdata[0]), nStates, &(thetakdata[0]), nParameters);
		thetakdata = mapper->map(thetakdata);

		thetak = mat(&(thetakdata[0]), nParameters, 1);
		xk = mat(&(xkdata[0]), nStates, 1);

		//	Perform observation
		(*H)(xk.memptr(), nStates, zkdata, nObservations);
		zk = mat(zkdata, nObservations, 1);
		//	Transform theta_k to assimilation values
		//	TODO Transform to data assimilation values

		Xk.col(i) = xk;
		Thetak.col(i) = thetak;
		Zk.col(i) = zk;
	}
	//	New state and its associated observation
	xk = mean(Xk, 1);
	thetak = mean(Thetak, 1);
	mat zkMean = mean(Zk, 1);	// Only constant alpha
	error = zkhat - zkMean;

	//	Update covariance matrixes
	LX = Xk * Dsigma;
	LTheta = Thetak * Dsigma;
	HL = Zk * Dsigma;
	U = Pa + HL.t() * (Wi * HL);

	invU = inv(U);

	//	Compute new estimate
	X = xk + LX * invU * HL.t() * (Wi * error);
	Theta = thetak + LTheta * invU * HL.t() * (Wi * error);

	delete[] zkdata;
	prevError = currError;
	currError = norm(error, 2);
	++currIt;

	return currError;
}

double MappedROUKF::executeStepParallel(vector<double> zkhatc, forwardOp A, observationOp H, int sigmaPoint, MPI_Comm world_comm, MPI_Comm sigmaMasters_comm) {

	//	Matrixes
	mat Thetak(nParameters, sigma.n_cols), Xk(nStates, sigma.n_cols), Zk(nObservations, sigma.n_cols);
	mat HL;
	mat invU = inv(U);
	mat C = chol(invU);

	//	Column vectors
	mat thetak(nParameters, 1);
	mat xk(nStates, 1);
	mat zk(nObservations, 1);
	mat zkhat(&(zkhatc[0]), nObservations, 1);

	double *zkdata;
	zkdata = new double[nObservations];

	//	Sampling
	mat s = sigma.col(sigmaPoint);
	xk = X + LX * C.t() * s;
	thetak = Theta + LTheta * C.t() * s;

	//	Propagate sigma point
	//	Copies xk to a primitive type in order to achieve armadillo independence in extern programs
	vector<double> thetakdata(thetak.memptr(), thetak.memptr() + nParameters);
	vector<double> xkdata(xk.memptr(), xk.memptr() + nStates);

	thetakdata = mapper->unmap(thetakdata);
	(*A)(&(xkdata[0]), nStates, &(thetakdata[0]), nParameters);
	thetakdata = mapper->map(thetakdata);

	thetak = mat(&(thetakdata[0]), nParameters, 1);
	xk = mat(&(xkdata[0]), nStates, 1);

	//	Perform observation
	(*H)(xk.memptr(), nStates, zkdata, nObservations);
	zk = mat(zkdata, nObservations, 1);

	if (sigmaMasters_comm != MPI_COMM_NULL) {
		//	Masters of each solver (rank < (nParameters + 1)) interchange data from executions
//		cout << "Masters gathering info ";
		MPI_Gather(xk.memptr(), nStates, MPI_DOUBLE, Xk.memptr(), nStates,
		MPI_DOUBLE, 0, sigmaMasters_comm);
//		cout << "xk ";
		MPI_Gather(thetak.memptr(), nParameters, MPI_DOUBLE, Thetak.memptr(), nParameters, MPI_DOUBLE, 0, sigmaMasters_comm);
//		cout << "thetak ";
		MPI_Gather(zk.memptr(), nObservations, MPI_DOUBLE, Zk.memptr(), nObservations, MPI_DOUBLE, 0, sigmaMasters_comm);
//		cout << "zk." << endl;
		;
	}

	//	Main master broadcasts the sigma points processing to all the workers in all solvers.
	MPI_Bcast(Xk.memptr(), (sigma.n_cols) * nStates, MPI_DOUBLE, 0, world_comm);
	MPI_Bcast(Thetak.memptr(), (sigma.n_cols) * nParameters, MPI_DOUBLE, 0, world_comm);
	MPI_Bcast(Zk.memptr(), (sigma.n_cols) * nObservations, MPI_DOUBLE, 0, world_comm);

	//	New state and its associated observation
	xk = mean(Xk, 1);
	thetak = mean(Thetak, 1);
	mat zkMean = mean(Zk, 1);	// Only for simplex case (other must use the weighted mean)
	error = zkhat - zkMean;

	//	Update covariance matrixes
	LX = Xk * Dsigma;
	LTheta = Thetak * Dsigma;
	HL = Zk * Dsigma;
	U = Pa + HL.t() * (Wi * HL);

	invU = inv(U);

	//	Compute new estimate
	X = xk + LX * invU * HL.t() * (Wi * error);
	Theta = thetak + LTheta * invU * HL.t() * (Wi * error);

	delete[] zkdata;

	prevError = currError;
	currError = norm(error, 2);
	++currIt;

	return currError;
}

void MappedROUKF::reset(int nObservations, int nStates, int nParameters, vector<double> observationsUncertainty, vector<double> parametersUncertainty, SigmaPointsGenerator::SIGMA_DISTRIBUTION sigmaDistribution) {
	this->nObservations = nObservations;
	this->nStates = nStates;
	this->nParameters = nParameters;

	X = zeros(nStates, 1);
	Theta = zeros(nParameters, 1);

	LTheta = eye(nParameters, nParameters);
	LX = zeros(nStates, nParameters);
	U = eye(nParameters, nParameters);
	Wi = eye(nObservations, nObservations);

	mat diagParameter(&(parametersUncertainty[0]), 1, nParameters);
	U.diag() = 1. / diagParameter;
	mat diagState(&(observationsUncertainty[0]), 1, nObservations);
	Wi.diag() = 1. / diagState;

	SigmaPointsGenerator::generateSigmaPoints(nParameters, sigmaDistribution, &sigma);
	alpha = 1. / sigma.n_cols;

	Dsigma = alpha * eye(sigma.n_cols, sigma.n_cols);
	Dsigma = Dsigma * sigma.t();
	Pa = sigma * Dsigma;

}

void MappedROUKF::getParameters(double** thetac) {
	vector<double> theta(Theta.memptr(), Theta.memptr() + nParameters);
	theta = mapper->unmap(theta);
	*thetac = new double[nParameters];
	memcpy(*thetac, &(theta[0]), nParameters * sizeof(double));
}

void MappedROUKF::setParameters(double *thetac) {
	vector<double> theta(thetac, thetac + nParameters);
	theta = mapper->map(theta);
	Theta = mat(&(theta[0]), nParameters, 1);
}

void MappedROUKF::replaceMapper(CompositeParameterMapper* mapper){
	if(this->mapper){
		//	Remap kalman parameters from previous kalman parameters space into the new one.
		vector<double> thetakdata(Theta.memptr(), Theta.memptr() + nParameters);
		thetakdata = this->mapper->unmap(thetakdata);
		delete this->mapper;
		thetakdata = mapper->map(thetakdata);
		Theta = mat(&(thetakdata[0]), nParameters, 1);
	}
	this->mapper = mapper;
}
